lib/decompress.ps1

```diff
@@ -1,123 +1,121 @@
 }
 
 function file_requires_7zip($fname) {
     $fname -match '\.((gz)|(tar)|(tgz)|(lzma)|(bz)|(bz2)|(7z)|(rar)|(iso)|(xz)|(lzh)|(nupkg))$'
 }
 
 function extract_7zip($path, $to, $recurse) {
-    $output = 7z x "$path" -o"$to" -y
-    if($lastexitcode -ne 0) { abort "Exit code was $lastexitcode." }
-
-    # check for tar
-    $tar = (split-path $path -leaf) -replace '\.[^\.]*$', ''
-    if($tar -match '\.tar$') {
-        if(test-path "$to\$tar") { extract_7zip "$to\$tar" $to $true }
+    $logfile = "$(Split-Path $path)\7zip.log"
+    &(file_path 7zip 7z.exe) x "$path" -o"$to" -y | Out-File "$logfile"
+    if ($lastexitcode -ne 0) {
+        abort "Failed to extract files from $path.`nLog file:`n  $(friendly_path $logfile)"
     }
-
-    if($recurse) { Remove-Item $path } # clean up intermediate files
+    if (Test-Path $logfile) {
+        Remove-Item $logfile -Force
+    }
+    if ((strip_ext (fname $path)) -match '\.tar$' -or (fname $path) -match '\.tgz$') {
+        $listfiles = &(file_path 7zip 7z.exe) l "$path"
+        if ($lastexitcode -eq 0) {
+            $tar = ([Regex]"(\S*.tar)$").Matches($listfiles[-3]).Value
+            extract_7zip "$to\$tar" $to $true
+        } else {
+            abort "Failed to list files in $path."
+        }
+    } # Check for tar
+    if ($recurse) {
+        Remove-Item $path -Force
+    } # Clean up compressed files
 }
 
-function extract_msi($path, $to) {
+function extract_msi($path, $to, $recurse) {
     $logfile = "$(split-path $path)\msi.log"
-    $ok = run 'msiexec' @('/a', "`"$path`"", '/qn', "TARGETDIR=`"$to`"", "/lwe `"$logfile`"")
-    if(!$ok) { abort "Failed to extract files from $path.`nLog file:`n  $(friendly_path $logfile)" }
-    if(test-path $logfile) { Remove-Item $logfile }
-}
-
-function lessmsi_config ($extract_dir) {
-    $extract_fn = 'extract_lessmsi'
-    if ($extract_dir) {
-        $extract_dir = join-path SourceDir $extract_dir
+    $useLessMsi = get_config MSIEXTRACT_USE_LESSMSI
+    if ($useLessMsi) {
+        &(file_path lessmsi lessmsi.exe) x "$path" "$to\" | Out-File "$logfile"
+        if ($lastexitcode -ne 0) {
+            abort "Failed to extract files from $path.`nLog file:`n  $(friendly_path $logfile)"
+        }
+        if (Test-Path "$to\SourceDir") {
+            movedir "$to\SourceDir" "$to" | Out-Null
+        }
     } else {
-        $extract_dir = "SourceDir"
+        $ok = run 'msiexec' @('/a', "`"$path`"", '/qn', "TARGETDIR=`"$to`"", "/lwe `"$logfile`"")
+        if (!$ok) {
+            abort "Failed to extract files from $path.`nLog file:`n  $(friendly_path $logfile)"
+        }
+        Remove-Item "$to\$(fname $path)"
     }
-
-    $extract_fn, $extract_dir
-}
-
-function extract_lessmsi($path, $to) {
-    Invoke-Expression "lessmsi x `"$path`" `"$to\`""
+    if (Test-Path $logfile) {
+        Remove-Item $logfile -Force
+    }
+    if ($recurse) {
+        Remove-Item $path -Force
+    } # Clean up compressed files
```

Comment:
Again comment on end of if branch. `if ($recurse) { # Clean up compressed files` looks much better and is more readable